ldap/ldap.go

```diff
@@ -0,0 +1,127 @@
+package ldap
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/jtblin/go-ldap-client"
+	ldapv2 "gopkg.in/ldap.v2"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+)
+
+//ValidateUser validates user by extracting creds from the request body, then
+//    authenticating against an ID backend and filling out the User struct therefrom
+func ValidateUser(req *http.Request, authConfig config.AuthConfig) (User, error) {
+
+	// access LDAP as the user we are validating
+	username, pwd, _ := req.BasicAuth()
+
+	log.Debugf("Connecting to LDAP with config: %v", authConfig)
+	bindDN := strings.Replace(authConfig.LDAP.LDAPBindDN, "{USERID}", username, -1)
+	log.Debugf("bindDN : %s", bindDN)
+
+	ldapAttributes := []string{authConfig.LDAP.LDAPMemberOfKey, authConfig.LDAP.LDAPUserIDKey, authConfig.LDAP.LDAPMailIDKey}
+	client := ldap.LDAPClient{
+		Base:               authConfig.LDAP.LDAPBase,
+		Host:               authConfig.LDAP.LDAPHost,
+		Port:               authConfig.LDAP.LDAPPort,
+		UseSSL:             authConfig.LDAP.LDAPUseSSL,
+		InsecureSkipVerify: authConfig.LDAP.LDAPInsecureSkipVerify,
+		BindDN:             bindDN,
+		BindPassword:       pwd,
+		UserFilter:         authConfig.LDAP.LDAPUserFilter,
+		GroupFilter:        authConfig.LDAP.LDAPGroupFilter,
+		Attributes:         ldapAttributes,
+	}
+	defer client.Close()
+
+	// uname and password (ID) correct?
+	authOk, err := isAuthenticated(&client, username, pwd)
+
+	if authOk { // get the rest of the info for this person -- groups, app ownershop etc.
```

Comment:
Similar to above, I think you want:  `if err != nil && authOk`.  Return the failure case first, then getUser.

Reply:
This is not correct. I will add comments to isAuthenticated to make this clear though. If you look, isAuthenticated may return false with an err == nil (a side effect of the call to ldapClient.Authenticate). So we can't assume anything based on err (i.e. things may have failed even if err is nil).

Reply:
OK, since authOk is a bool it's safe as is.